Samples/MapScenarios/MapScenarios/Scenarios/WalkTime/WalkTimeViewController.swift

```diff
@@ -0,0 +1,177 @@
+            return -1
+        }
+        
+        var distance: CLLocationDistance = 0
+        
+        // The distance for rest of instructions
+        for instruction in restInstructions {
+            distance += instruction.distance
+        }
+        
+        // The distance from current location to the end of current instruction
+        if let instructionEndLocation = currentInstruction.points.last?.coordinate {
+            let userLocation = CLLocation(latitude: lastUpdateLocation.coordinate.latitude, longitude: lastUpdateLocation.coordinate.longitude);
+            let endLocation = CLLocation(latitude: instructionEndLocation.latitude, longitude: instructionEndLocation.longitude)
+            distance += userLocation.distance(from: endLocation)
+        }
+        
+        return distance
+    }
+}
+
+// MARK: - PWMapViewDelegate
+
+extension WalkTimeViewController: PWMapViewDelegate {
+    
+    func mapView(_ mapView: PWMapView!, locationManager: PWLocationManager!, didUpdateIndoorUserLocation userLocation: PWUserLocation!) {
+        lastUpdateLocation = userLocation
+    }
+    
+    func mapViewStartedSnappingLocation(toRoute mapView: PWMapView!) {
+        snappedLocation = true
+    }
+    
+    func mapViewStoppedSnappingLocation(toRoute mapView: PWMapView!) {
+        snappedLocation = false
+    }
+    
+    func mapView(_ mapView: PWMapView!, didChange instruction: PWRouteInstruction!) {
+        // Update walk time when blue dot is not acquired
+        if lastUpdateLocation == nil, let currentIndex = self.mapView.currentRoute.routeInstructions.firstIndex(of: instruction), let restInstructions = self.mapView.currentRoute.routeInstructions?[currentIndex...] {
+            var distance: Double = 0
+            for instruction in restInstructions {
+                distance += instruction.distance
+            }
+            self.walkTimeView?.updateWalkTime(distance: distance, averageSpeed: 0)
+        }
+    }
+}
+
+// MARK: - CLLocationManagerDelegate
+
+extension WalkTimeViewController: CLLocationManagerDelegate {
+    
+    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
+        switch status {
+        case .authorizedAlways, .authorizedWhenInUse:
+            // Re-register location manager
+            mapView.unregisterLocationManager()
+            let managedLocationManager = PWManagedLocationManager(buildingId: buildingIdentifier)
+            DispatchQueue.main.async { [weak self] in
+                self?.mapView.register(managedLocationManager)
+            }
+        default:
+            print("Not authorized to start PWManagedLocationManager")
+        }
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+        if let location = locations.first {
+            speedArray.append(location.speed)
```

Comment:
is this location.speed accurate? my concern is since your GPS dot might be barely moving inside, or just overall inaccurately, it will be way off

Reply:
Right, that's why it's to average last couple of reported speed as user's current speed. Will take a look at your solution.

Reply:
I imagine if a user is taking a look at the route before starting to walk, they will have a speed of near 0, which would make the time estimate really high

Reply:
Couldn't we use a constant average person walking speed until we get some speed data coming in?

Reply:
First, the location updates usually starting before the indoor map and route is fully loaded.
Second, the locaitonManager.activityType is setting as `.fitness`, then the updated speed is acceptable.
Third, there's a reasonable walk speed in the WalkTimeView, it will be replaced with the default walk speed if it's out of the range.